File: lc_cpp/lcCpp/greedy/maxFreeTime3439.cpp
//
// Created by 24276 on 2025/7/10.
//

#define V vector
#define pb push_back
#include <iostream>
#include <vector>
using namespace std;
class max_free_time3439 {
public:
    int maxFreeTime(int eventTime, int k, vector<int>& startTime, vector<int>& endTime) {
        // 滑动窗口 + 贪心  n个会议有n+1个空闲时间 ps: 左右
        int n = static_cast<int>(startTime.size());
        vector<int> free_t(n + 1);
        free_t[0] = startTime[0];   //存储两个会议之间的空闲间隔
        for(int i = 1;i < n;i++) {
            free_t[i] = startTime[i] - endTime[i - 1];
        }
        free_t[n] = eventTime - endTime[n - 1];
        int ans = 0, s = 0;
        for (int i = 0;i <= n;i++) {
            s += free_t[i];
            if(i < k) {
                continue;
            }
            ans = max(ans,s);
            s -= free_t[i - k];
        }
        return ans;
    }
};

int main() {
    max_free_time3439 ps;
    int eventTime = 5,k = 1;
    vector<int> startTime = {1,3},endTime = {2,5};
    std::cout << ps.maxFreeTime(eventTime,k,startTime,endTime);
    return 0;
}

File: lc_py/lcPy/sliding_window/max_free_time3439.py
from typing import List


class Solution:
    def maxFreeTime(self, eventTime: int, k: int, startTime: List[int], endTime: List[int]) -> int:
        '''
        重新安排会议 不要有交集 不要改变时长
        '''
        n = len(startTime)
        # n个会议 n+1 个空闲
        free = [0] * (n + 1)

        free[0] = startTime[0]  # 最左侧 ps:startTime = 0?

        for i in range(1,n):
            free[i] = startTime[i] - endTime[i - 1]
        free[n] = eventTime - endTime[-1]

        ans = s = 0

        for i,f in enumerate(free):
            s += f
            if i < k:
                continue
            ans = max(ans,s)
            s -= free[i - k]
        return ans


if __name__ == '__main__':
    s = Solution()
    print(s.maxFreeTime(5, 1, [1, 3], [2, 5]))
    